Allow contents of Paragraph to be "connected" to a meta custom field

michalczaplinski opened the pull request:
Following up on work in #51375

This PR is a counterpart to **and** stacked on top of #53241 because we need some code from there to make it work. So #53241  should be merged first.

## What does this do?

We're adding a way to "connect" the contents of the Paragraph block to a meta custom field.


https://github.com/WordPress/gutenberg/assets/5417266/8f9ddba1-f4e5-4914-b236-422fc7f4c26b

cbravobernal committed:
Add custom fields experimental setting

cbravobernal committed:
Add connections experimental setting, block inspector control and block attributes

michalczaplinski committed:
Add "connections" block attribute and get the meta field

michalczaplinski committed:
Merge branch 'experiment/add-custom-connectinons-inspector-control' into paragraph-block-custom-fields

michalczaplinski committed:
Get rid of a phpcs warning

michalczaplinski committed:
Update attribute acces and source definition

Changed the attribute access from `$block_type->attributes` to `$block['attrs']`. Also updated attribute 'source' value from 'meta' to 'meta_fields'. These changes have been applied in both `blocks.php` and block attributes declaration in `block.json`.

michalczaplinski committed:
Updated the 'connections' property to include 'attributes'

michalczaplinski committed:
Merge branch 'experiment/add-custom-connectinons-inspector-control' into paragraph-block-custom-fields

michalczaplinski committed:
Add `attributes` to TextControl value

michalczaplinski committed:
Add `attributes` to TextControl value

michalczaplinski committed:
Wrap the PHP code in experimental flag

michalczaplinski committed:
Merge branch 'experiment/add-custom-connectinons-inspector-control' into paragraph-block-custom-fields

michalczaplinski committed:
Update parens

michalczaplinski committed:
Return early if there is no support for connections

michalczaplinski committed:
Updated block type checks

Rearranged the block type checks. Now, the code first checks if a given block is in the accepted types list (Paragraph and Image) and then it checks if the block type is null. This is a slight revamp of the existing checks for block support and block type acceptance.

michalczaplinski committed:
Updated experimental blocks logic

- Removed unused test functions.
- Update the comment above the tag name query in the `custom-sources` to clarify future improvements for
supporting more block types.

michalczaplinski committed:
Add a comment in meta.php

cbravobernal committed:
Update comment

cbravobernal committed:
Clear the paragraph if meta value is cleared

cbravobernal committed:
Use placeholders instead of content

cbravobernal commented on lib/experimental/custom-sources/meta.php:
```diff
@@ -0,0 +1,36 @@
+		// We should probably also check if the meta field exists but for now it's okay because
+		// if it doesn't, `get_post_meta()` will just return an empty string.
+		$meta_value = get_post_meta( $block_instance->context['postId'], $meta_field, true );
+		$p          = new WP_HTML_Tag_Processor( $block_content );
```
`$p` and `$p2` are not quite descriptive 😅

michalczaplinski replied:
true 😅 I'll change that

cbravobernal commented on lib/experimental/custom-sources/meta.php:
```diff
@@ -0,0 +1,36 @@
+			$p2->set_attribute( $name, $p->get_attribute( $name ) );
+		}
+
+		return $p2 . '';
```
Do we need `get_updated_html()` or `apply_attributes_updates()` here?

michalczaplinski replied:
That's a good catch! I've just double checked and we don't need the `get_updated_html()` because `WP_HTML_Tag_Processor::__toString()` calls `get_updated_html()` automatically ([link](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/__tostring/#source)).

I don't know if we need `apply_attributes_updates()` because I don't understand its use case. 

@dmsnell Could you enlighten me on what `apply_attributes_updates()` does? Is it always necessary to call it after calling `p->set_attribute()`? Or it only necessary in some circumstances?

dmsnell replied:
userspace should never worry about calling `apply_attribute_updates()`. that should be an internal-use only function

michalczaplinski replied:
gotcha. Thanks @dmsnell !

michalczaplinski committed:
Update names of HTML processor instances in meta.php

michalczaplinski committed:
Add extra comments

michalczaplinski committed:
Merge branch 'experiment/add-custom-connectinons-inspector-control' into paragraph-block-custom-fields

michalczaplinski committed:
Rephrase the placeholder comment

michalczaplinski committed:
Check if current block is a paragraph or image

michalczaplinski committed:
Abstract the attribute name to use for the connection

dmsnell commented on lib/experimental/custom-sources/meta.php:
```diff
@@ -0,0 +1,36 @@
+			$updated_tags->set_attribute( $name, $tags->get_attribute( $name ) );
+		}
+
+		return $updated_tags->get_updated_html() . '';
```
curious what the ` . ''` is here for. were you running into errors that you're trying to cover?

michalczaplinski replied:
Good catch. I hate to say this, but I'm not sure. Either it was something that Copilot suggested to me or I copied this from Riad's PR. Either way, I'll remove it because we don't need it, thanks!

SantosGuillamot commented on packages/block-library/src/paragraph/block.json:
```diff
@@ -28,6 +29,14 @@
 		"direction": {
 			"type": "string",
 			"enum": [ "ltr", "rtl" ]
+		},
+		"connections": {
+			"type": "object",
+			"attributes": {
+				"content": {
+					"source": "meta_fields"
+				}
+			}
```
Do we need to define the "connections" attribute? Ideally, I would like to explore the possibility of being able to populate that block attribute without having to define it in the `block.json`.  For example, that's what's happening with styles, right? Maybe it makes more sense to just have `supports.connections: true`?

michalczaplinski replied:
Yeah, that makes sense Mario, let me make that change.

michalczaplinski replied:
done in 367ee0c34a482c1199031f3768746c5fdcd12304

SantosGuillamot commented on lib/experimental/blocks.php:
```diff
@@ -121,3 +121,69 @@ function gutenberg_register_metadata_attribute( $args ) {
+		$meta_custom_source = require __DIR__ . '/custom-sources/meta.php';
+		$block_type         = $block_instance->block_type;
+
+		// Whitelist of the block types that support custom sources
+		// Currently, we only allow the Paragraph and Image blocks to use custom sources.
+		if ( ! in_array( $block['blockName'], array( 'core/paragraph', 'core/image' ), true ) ) {
+			return $block_content;
+		}
```
Would it make sense to create a whitelist of block attributes instead of just blocks? Both in the paragraph and image, there are block attributes that won't work with the connections.

michalczaplinski replied:
This serves as a "whitelist" of the block attributes:

https://github.com/WordPress/gutenberg/blob/f0bcb0fb04f9412fb67303f12c101c073a7bc0c6/packages/block-editor/src/hooks/custom-fields.js#L67-L68

Do you think that's sufficient?

michalczaplinski replied:
I suppose we could only check the `content` attribute of the Paragraph and the `url` of the Image, instead of iterating over the attributes on this line: 

https://github.com/WordPress/gutenberg/blob/60fa0613981cbd1cc45010cefdae40d4a49be509/lib/experimental/blocks.php#L161

SantosGuillamot replied:
I don't know what the best mechanism is, and I guess we will have to iterate it, but right now if I add a connection manually to the block (without using the Block Editor), I could include any block attribute and the PHP logic will try to replace that, right?

Apart from that, keep in mind that in the future we will probably include support for other Image block attributes like `title`. I would say that the easier it is to add block attributes to those lists the better. As we will be adding support for them progressively.

michalczaplinski replied:
> I don't know what the best mechanism is, and I guess we will have to iterate it, but right now if I add a connection manually to the block (without using the Block Editor), I could include any block attribute and the PHP logic will try to replace that, right?

Yep, that's true. OK, makes sense that we create an explicit, more centralized whitelist in the `blocks.php` file.

michalczaplinski replied:
done in 49d5cb6e3c6a167825af85cf00e40370cf767988

SantosGuillamot commented on lib/experimental/blocks.php:
```diff
@@ -121,3 +121,69 @@ function gutenberg_register_metadata_attribute( $args ) {
+	 * @param array    $block Block attributes.
+	 * @param WP_Block $block_instance The block instance.
+	 */
+	function gutenberg_render_custom_sources( $block_content, $block, $block_instance ) {
+		$meta_custom_source = require __DIR__ . '/custom-sources/meta.php';
```
I'm a bit confused about the "custom sources" naming. Is this what we are doing? Are we rendering custom sources? If I understood it correctly, we are:

* Processing the connections block attribute.
* Getting the value from the connections source. Meta in this case using `get_post_meta`.
* Replacing the HTML according to the block attributes source. `html` in the case of the paragraph.

In my opinion, this last step should be independent of the connections source. If I get the value from a different source, I would still have to replace the same HTML, but with a different value. Is this right?

Just to have more context, I believe in the future this distinction will be more relevant:

* **Connections sources**: From where we get the value. Right now it is only meta, but it could be drupal, parent-block, my-custom-database...
* **Block attributes source**: Dictates which part of the HTML needs to be replaced. Thanks to the block attributes source, it's possible to know the HTML we have to replace. If it is `html` we will replace the inner content. If it is `attribute` we will replace an HTML attribute.

That's why it makes sense to separate the code as well.

michalczaplinski replied:
> * Processing the connections block attribute.
> * Getting the value from the connections source. Meta in this case using `get_post_meta`.
> * Replacing the HTML according to the block attributes source. `html` in the case of the paragraph.

You're right, that's the right abstraction of the 3 steps that we perform. As it stands right now, the 2nd & 3rd step are not separated. It makes sense that they should be.

Also, some of the naming is a direct copy from Riad's PR (https://github.com/WordPress/gutenberg/pull/51375) which is not 100% accurate anymore. 

I ll change it.

michalczaplinski replied:
The code should be better abstracted now to reflect those 3 separate steps. 1 & 3 are both inside of the same function but we can abstract it further when we add support for the Image block & other attributes. Then we can abstract the part that replaces the different attribute sources (`html`, `attribute`, etc.).

We should discuss consistent naming though of Step 2. Should it be:
- custom sources
- custom connections
- connection sources
- custom connection sources
- ...something else?

SantosGuillamot replied:
Yes, I believe the whole naming should be discussed. Another option I was thinking of was using "Dynamic data" instead of "Connections". So it could be "Dynamic data sources". But I didn't think enough about it. What doesn't quite fit me is the use of "custom". What does "custom" mean in this context? Which sources are not custom?

Do we need it for this initial experiment? Is it too difficult to change it afterwards?

michalczaplinski replied:
> Is it too difficult to change it afterwards?
Nope, it won't be difficult at all, I just wanted to start this conversation and have consistent naming even if it gets changed later 🙂 .

I guess "custom" means something like "different than whatever WP uses to get the content normally" but I agree that it's a bit meaningless.

I like "Connections" a little bit better than "Dynamic Data" because the latter can mean many things in many different contexts. Most data is dynamic rather than static in programming 🤓. I think "Connection Sources" sounds best so far. "Connections" alone is also a bit vague. "Connection Sources" implies "_data_ sources" which is roughly what we are referring to here.

michalczaplinski replied:
Maybe "block connections" is also a good name? We're **connecting stuff to blocks** after all.

SantosGuillamot replied:
Yep, I was thinking about names using this sentence: **we are connecting block attributes to values that might change depending on the context (post_id, parent block…)**. But not sure yet which is the best abstraction 😄

michalczaplinski replied:
Hmm, I would just go with "block connections" as a name for the whole thing (like I use it in https://github.com/WordPress/gutenberg/pull/53247/files#diff-be8a354ce9c01e85c0fdf04c75d76880cdc6c81992db7a86a7a0515dc11a9833R135) and "connection sources" for the thing that does the connecting to meta fields/parent blocks/other DBs, etc.

We can always change it in the future.

If you don't object, then I ll go ahead and change it.

SantosGuillamot replied:
Sure! We can always change the naming later if we agree on a different one.

SantosGuillamot reviewed (commented):
I left some comments to discuss. Apart from that, I believe there are some styles broken.

michalczaplinski committed:
rename `connections` to `__experimentalConnections`

michalczaplinski committed:
Do not export `addAttribute()` or `withInspectorControl()`

michalczaplinski committed:
Merge branch 'experiment/add-custom-connectinons-inspector-control' into paragraph-block-custom-fields

michalczaplinski committed:
Renamed '__experimentalConnections' to 'connections' in block settings

michalczaplinski committed:
Renamed '__experimentalConnections' to 'connections' in block settings

michalczaplinski committed:
Renamed '__experimentalConnections' to 'connections' in block settings

michalczaplinski committed:
Remove `connections` attribute from block.json

michalczaplinski committed:
Remove `get_updated_html() . ''` in meta.php

michalczaplinski committed:
Add an allowlist of blocks/attributes

michalczaplinski committed:
Refactor blocks.php & custom sources

- Separate the "connections" into own file.
- Move the code that renders the HTML using the meta fields back into blocks.php

michalczaplinski committed:
Remove trailing comma?

michalczaplinski commented on packages/block-editor/src/hooks/custom-fields.js:
```diff
@@ -14,18 +14,18 @@ import { InspectorControls } from '../components';
- * Filters registered block settings, extending attributes to include `__experimentalConnections`.
+ * Filters registered block settings, extending attributes to include `connections`.
  *
  * @param {Object} settings Original block settings.
  *
  * @return {Object} Filtered block settings.
  */
 function addAttribute( settings ) {
 	if ( hasBlockSupport( settings, '__experimentalConnections', true ) ) {
-		// Gracefully handle if settings.attributes is undefined.
+		// Gracefully handle if settings.attributes.connections is undefined.
 		settings.attributes = {
 			...settings.attributes,
-			__experimentalConnections: {
+			connections: {
```
I'm not sure why these changes show up in GithHub's UI. They are already in the other PR: https://github.com/WordPress/gutenberg/pull/53241

SantosGuillamot replied:
Maybe we need to rebase/merge the other PR branch. If I am not mistaken the last merge was done before the change you mention.

michalczaplinski replied:
yep, I had to merge `experiment/add-custom-connectinons-inspector-control` into this one before pushing again 🙂

SantosGuillamot reviewed (approved):
Looks good to me for this first iteration 🙂

michalczaplinski committed:
Update naming:
- "custom sources" -> connection sources
- use "block connections" consistently

michalczaplinski commented:
Great, I've just updated the naming to consistently use:
- "connection sources" as the thing doing the connecting to meta fields/other sources in the future
- "block connections" as the name for the whole feature.

I'll first merge #53241 and then this one.

michalczaplinski committed:
Merge branch 'trunk' into experiment/add-custom-connectinons-inspector-control

michalczaplinski committed:
Merge branch 'experiment/add-custom-connectinons-inspector-control' into paragraph-block-custom-fields

youknowriad commented on lib/experimental/blocks.php:
```diff
@@ -121,3 +121,107 @@ function gutenberg_register_metadata_attribute( $args ) {
+		$blocks_attributes_allowlist = array(
+			'core/paragraph' => array( 'content' ),
+			'core/image'     => array( 'url' ),
+		);
```
Can we retrieve this from a block support instead of a hardcoded list.

youknowriad replied:
Actually it's the user that should define which attribute to connect, so this is coming from the block attribute value  normally no?

michalczaplinski replied:
See my comment in https://github.com/WordPress/gutenberg/pull/53241#discussion_r1293545604 :) 

Yes, but for now we want to limit this to core blocks and those two attributes. We'll expand it in the future. 

So far it's not well-defined how users are going to opt-in to connecting their custom blocks' attributes to custom fields.

youknowriad commented on lib/experimental/blocks.php:
```diff
@@ -121,3 +121,107 @@ function gutenberg_register_metadata_attribute( $args ) {
+			}
+
+			// Get the content from the connection source.
+			$custom_value = $connection_sources[ $attribute_value['source'] ](
```
This is weird, I'd have preferred a unique name for all connection, like `get_value` or `apply` or something.

michalczaplinski replied:
Oh, maybe the confusion is coming from this line: https://github.com/WordPress/gutenberg/blob/676f25961bfab695bacb5dd7ebb10443e49bce23/lib/experimental/connection-sources/index.php#L9

I should have removed the `"name"` because it serves no purpose at the moment.

So in your PR, there was an assumption that there's one file per connection source (`attributes-custom-sources/meta.php`, `attributes-custom-sources/some-other-source.php`) and here they are all in under keys of one array. I say "all" but there is only one type of connection source, anyway 😅 .

youknowriad replied:
ah ok, I guess depending on the complexity we could split but agree, seems like implementation detail for now but let's aim for the simplest. I do expect us to allow registering sources at some point so I think "one config array per source" is probably better ultimately (as opposed to all configs in a single array).

youknowriad commented on packages/block-library/src/paragraph/block.json:
```diff
@@ -7,6 +7,7 @@
 	"description": "Start with the basic building block of all narrative.",
 	"keywords": [ "text" ],
 	"textdomain": "default",
+	"usesContext": [ "postId" ],
```
I wonder whose responsibility is it to do that. In other words, the block is not using the "postId" explicitly, so why add the "useContext" here. The "connection framework" should inject/filter this somehow.

michalczaplinski replied:
Do you mean that in the context of Core blocks or custom blocks or both?

youknowriad replied:
both

michalczaplinski replied:
OK! For now, I think that for core blocks this is fine. But you're right that for custom blocks we'll have to come up with a proper mechanism here. Are you all right with that?

youknowriad replied:
Yeah I think it's part of the definition of a "custom source" to say which change it needs to make to block registration. I think we should use it for both core and third-party blocks but yeah it's fine to start without it.

gziolo replied:
For now, you could update `usesContext` and pass `postId` in PHP when processing `gutenberg_render_block_connections`. This way, it won't get unnecessarily wired for all Paragraph block instances through `block.json` definition.

SantosGuillamot replied:
Even though this is the first implementation and custom blocks are not supported yet, I believe it'd be better to start creating mechanisms that would work for both if possible. Mainly because we already know we will want to support that in the future.

talldan replied:
Coming in a bit late here.

The current work on partial syncing (#54233) is using the store actions/selectors for attributes to inject/update values, but it might be difficult to pull this off if block context is being used for custom fields, as I believe it uses react context in JS.

While they're still proofs of concepts, the current work for custom fields/partial syncing does need to remain compatible, so it might be good to explore ways to read/write custom field values in the editor soon though so that we have a clear idea of the problems that need to be solved.

mcsf commented on lib/experimental/blocks.php:
```diff
@@ -121,3 +121,107 @@ function gutenberg_register_metadata_attribute( $args ) {
+				return $block_content;
+			};
+			$tag_name     = $tags->get_tag();
+			$markup       = "<$tag_name>$custom_value</$tag_name>";
```
This line is scary. Any plans for escaping? It's not trivial, since the goal is to be able to inject rich text and not just simple values.

dmsnell replied:
thanks for catching this @mcsf. this definitely should be `"<$tag_name>" . esc_html( $custom_value ) . "</$tag_name>"` for now.

actually @michalczaplinski let's talk about how to sub-class the Tag Processor so this is cleaner all over

SantosGuillamot replied:
Thanks for raising it! We should definitely do that. Keep in mind that this code was considered experimental and we'll need to refactor it.

mcsf replied:
Definitely, but all too often lines of experimental code somehow make their way into production code. :) I'd really patch ASAP, even if just with `esc_html`.

michalczaplinski replied:
Thanks for catching this! 

I want to take a holistic look at all the code shipped behind the experimental flag. It's very likely most of it will be replaced altogether. The team is on support rotation this week so it'll have to wait till next week at least 🙂 

@dmsnell yes, let's talk!
